lib/improver/nbhood.py

```diff
@@ -55,107 +59,243 @@ class BasicNeighbourhoodProcessing(object):
     # Max extent of kernel in grid cells.
     MAX_KERNEL_CELL_RADIUS = 500
 
-    def __init__(self, radius_in_km, unweighted_mode=False):
+    def __init__(self, radii_in_km, lead_times=None, unweighted_mode=False):
         """
         Create a neighbourhood processing plugin that applies a smoothing
         kernel to points in a cube.
 
         Parameters
         ----------
 
-        radius_in_km : float
-            The radius in kilometres of the neighbourhood kernel to
+        radii_in_km : float or List (if defining lead times)
+            The radii in kilometres of the neighbourhood kernel to
             apply. Rounded up to convert into integer number of grid
             points east and north, based on the characteristic spacing
             at the zero indices of the cube projection-x/y coords.
-
+        lead_times : None or List
+            List of lead times or forecast periods, at which the radii
+            within radii_in_km are defined.
         unweighted_mode : boolean
             If True, use a circle with constant weighting.
             If False, use a circle for neighbourhood kernel with
             weighting decreasing with radius.
 
         """
-        self.radius_in_km = float(radius_in_km)
+        if isinstance(radii_in_km, list):
+            self.radii_in_km = map(float, radii_in_km)
+        else:
+            self.radii_in_km = float(radii_in_km)
+        self.lead_times = lead_times
+        if self.lead_times is not None:
+            if len(radii_in_km) != len(lead_times):
+                msg = ("There is a mismatch in the number of radii "
+                       "and the number of lead times. "
+                       "Unable to continue due to mismatch.")
+                raise ValueError(msg)
         self.unweighted_mode = bool(unweighted_mode)
 
     def __str__(self):
-        result = ('<NeighbourhoodProcessing: radius_in_km: {};' +
+        result = ('<NeighbourhoodProcessing: radii_in_km: {};' +
                   'unweighted_mode: {}>')
         return result.format(
-            self.radius_in_km, self.unweighted_mode)
+            self.radii_in_km, self.unweighted_mode)
+
+    @staticmethod
+    def _find_required_lead_times(cube):
+        """
+        Determine the lead times within a cube, either by reading the
+        forecast_period coordinate, or by calculating the difference between
```

Comment:
That may still leave an ambiguity. A 2-hour period centred around a particular time may require a different neighbourhood size from a 6-hour period around the same time. That's because the longer time window is typically a smoother field that has filtered out more temporal and spatial variability. Having said that, we need to be very careful to make sure that if we have two different neighbourhood sizes for the same central time it doesn't generate contradictions, and that needs some thought - so the concept of a "most valid time" (halfway between) is definitely safer - certainly good to start with, but I wouldn't want to rule out the possibility of being able to define by two time definitions at some point.